Document using multiple scheduler profiles (#19172)
kubernetes/website: content/en/docs/concepts/scheduling/kube-scheduler.md

If there is more than one node with equal scores, kube-scheduler selects
one of these at random.

There are two supported ways to configure the filtering and scoring behavior
of the scheduler:

1. [Scheduling Policies](/docs/reference/kube-scheduler/policies)

> 1. [Scheduling Profiles](/docs/reference/kube-scheduler/profiles)

Reviewer: A nit:
```suggestion
1. [Scheduling Policies](/docs/reference/kube-scheduler/policies/)
1. [Scheduling Profiles](/docs/reference/kube-scheduler/profiles/)
```

also, I recommend adding more text here. Someone reading the  concept page should learn at least a summary of difference between Scheduling Policies and Scheduling Profiles, even if they do not see the full lists of predicates and priorities that are applied.

If there is more than one node with equal scores, kube-scheduler selects
one of these at random.

There are two supported ways to configure the filtering and scoring behavior
of the scheduler:

1. [Scheduling Policies](/docs/reference/kube-scheduler/policies) allows you to
  configure *Predicates* for filtering and *Priorities* for scoring.
1. [Scheduling Profiles](/docs/reference/kube-scheduler/profiles) allows you to
  configure Plugins that implement different scheduling stages, including:
  `QueueSort`, `Filter`, `Score`, `Bind`, `Reserve`, `Permit` and others. You

>   can also configure the kube-scheduler to run different profiles.

Reviewer: Fixing broken hyperlinks; also revising for grammar and to match website style guide.
```suggestion
1. [Scheduling Policies](/docs/reference/scheduling/policies/) allow you to
  configure _predicates_ for filtering and _priorities_ for scoring.
1. [Scheduling Profiles](/docs/reference/scheduling/profiles/) allow you to
  configure Plugins that implement different scheduling stages, including:
  `QueueSort`, `Filter`, `Score`, `Bind`, `Reserve`, `Permit`, and others. You
  can also configure the kube-scheduler to run different profiles.
```

If there is more than one node with equal scores, kube-scheduler selects
one of these at random.

There are two supported ways to configure the filtering and scoring behavior
of the scheduler:

> 1. [Scheduling Policies](#scheduling-policies),

Reviewer 2: ```suggestion
1. [Scheduling Policies](#scheduling-policies)
```

> 1. [Scheduling Profiles](#scheduling-profiles).

Reviewer 2: ```suggestion
1. [Scheduling Profiles](#scheduling-profiles)
```

If there is more than one node with equal scores, kube-scheduler selects
one of these at random.

> The are two supported ways to configure the filtering and scoring behavior

Author: Done

of the scheduler:

> 1. Scheduling Policies and

Author: Done

1. Scheduling Profiles.

### Scheduling policies

A scheduling Policy can be used to specify the predicates and priorities that
the scheduler runs to filter and score nodes, respectively.

A policy can be specified through a file (via `--policy-config-file` flag) or a
ConfigMap (via `--policy-configmap` flag) that follows the
[Policy type](https://pkg.go.dev/k8s.io/kube-scheduler@v0.18.0/config/v1?tab=doc#Policy).

> ### Filtering

Reviewer 3: I think we should move the section `Filtering` and `Scoring` in section "Scheduling with kube-scheduler", right before introducing scheduler policies and profiles, like this:

## Scheduling with kube-scheduler

...

### FIltering

### Scoring

## Configuring kube-scheduler (or similar name)

### Scheduling policies

### Scheduling profiles

Author: I don't think so. Because those sections use the names (for Predicates and Priorities) that only the Policy API exposes.

Reviewer 3: Or, we put a table for mapping Predicate/Priority name to Plugin name. Then guide the audience which literals to use upon different configuring manners (policy vs. profile). WDYT?

Author: The only caveat is that there is no direct mapping, because a predicate plus a priority might map to a single plugin.

What about using the subtitles `Predicates` and `Priorities` inside the `Scheduling Policies` section?

Reviewer 3: Version https://5e66a3ce506312000a06f41d--kubernetes-io-vnext-staging.netlify.com/docs/concepts/scheduling looks good to me.

  This applies for both bound and unbound
  {{< glossary_tooltip text="PVCs" term_id="persistent-volume-claim" >}}.

#### Scoring

> The following priorities implement scoring:

Reviewer: Is “priorities” the right word here?

Author: Yes. `Predicates` and `Priorities` are the names of the fields in the Policy to configure filtering and scoring, respectively.

Reviewer: Is that name is from alpha code? If so, it's fixable and I might well (separately) suggest changing it: the English noun “priority” doesn't suggest _thing that determines part of an overall score for each of a set of items_. (Compare “filter” and “predicate”, which both suggest _thing that determines which items from a set are retained_).

Meanwhile, I wonder if there's a way to avoid introducing wording that might confuse readers.
I will try to suggest different wording.

Author: No, Policy is a v1 API :(
That's why our new v1alpha2 Profile API uses the term scoring instead.

Reviewer: Ah, makes sense.

Reviewer: Consider making priorities italic, to indicate that you're introducing a concept.

  that don't have Pods for the service already assigned there. The overall outcome is
  that the Service becomes more resilient to a single Node failure.

- `EqualPriority`: Gives an equal weight of one to all nodes.

- `EvenPodsSpreadPriority`: Implements preferred
  [pod topology spread constraints](/docs/concepts/workloads/pods/pod-topology-spread-constraints/).

### Scheduling profiles

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

Scheduling profiles allow more flexibility when configuring a scheduler.
In a profile, Plugins that implement scheduling behaviors can be enabled,
disabled and reordered. A plugin implements one or more scheduling extension
points, namely:

1. **Queue sort**: These plugins provide an ordering function that is used to
   sort pending Pods in the scheduling queue. Exactly one queue sort plugin
   may be enabled at a time.
1. **Pre-filter**: These plugins are used to pre-process or check information
   about a Pod or the cluster before filtering.
1. **Filter**: These plugins are the equivalent of Predicates in a scheduling
   Policy and are used to filter out nodes that can not run the Pod. Filters
   are called in the configured order.

> 1. **Pre-score**: This is an informational extension point that can be used

Reviewer 2: ```suggestion
1. **PreScore**: This is an informational extension point that can be used
```

   for doing pre-scoring work.
1. **Score**: These plugins provide a score to each node that has passed the
   filtering phase. The scheduler will then select the node with the highest
   weighted scores sum.
1. **Reserve**: This is an informational extension point that notifies plugins
   when resources have being reserved for a given Pod.
1. **Permit**: These plugins can prevent or delay the binding of a Pod.

> 1. **Pre-bind**: These plugins perform any work required before a Pod is bound.

Reviewer 2: ```suggestion
1. **PreBind**: These plugins perform any work required before a Pod is bound.
```

1. **Bind**: The plugins bind a Pod to a Node. Bind plugins are called in order
   and once one has done the binding, the remaining plugins are skipped. At
   least one bind plugin is required.

> 1. **Post-bind**: This is an informational extension point that is called after

Reviewer 2: ```suggestion
1. **PostBind**: This is an informational extension point that is called after
```

   a Pod has been bound.

> 1. **Un-reserve**: This is an informational extension point that is called if

Reviewer 2: ```suggestion
1. **Unreserve**: This is an informational extension point that is called if
```

   a Pod is rejected after being reserved and put on hold by a Permit plugin.

The following plugins, enabled by default, implement one or more of these
extension points:

- `DefaultTopologySpread`: Favors spreading across nodes for Pods that belong to
  services, replica sets and stateful sets.

>   Extension points: *Pre-score*, *Score*.

Reviewer 2: ```suggestion
  Extension points: *PreScore*, *Score*.
```

- `ImageLocality`: Favors nodes that already have the container images that the
  Pod runs.
  Extension points: *Score*.
- `TaintToleration`: Implements
  [taints and tolerations](/docs/concepts/configuration/taint-and-toleration/).
  Implements extension points: *Filter*, *Pre-score*, *Score*.
- `NodeName`: Checks if a Pod spec node name matches the current node.
  Extension points: *Filter*.
- `NodePorts`: Checks if a node has free ports for the requested Pod ports.

>   Extension points: *Pre-filter*, *Filter*.

Reviewer 2: ```suggestion
  Extension points: *PreFilter*, *Filter*.
```

- `NodePreferAvoidPods`: Scores nodes according to the node annotation
  `scheduler.alpha.kubernetes.io/preferAvoidPods`.
  Extension points: *Score*.
- `NodeAffinity`: Implements
  [node selectors](/docs/concepts/configuration/assign-pod-node/#nodeselector)
  and [node affinity](/docs/concepts/configuration/assign-pod-node/#node-affinity).
  Extension points: *Filter*, *Score*.
- `PodTopologySpread`: Implements
  [Pod topology spread](/docs/concepts/workloads/pods/pod-topology-spread-constraints/).

>   Extension points: *Prefilter*, *Filter*, *Pre-score*, *Score*.

Reviewer 2: ```suggestion
  Extension points: *PreFilter*, *Filter*, *PreScore*, *Score*.
```

- `NodeUnschedulable`: Filters out nodes that have `.spec.unschedulable` set to
  true.
  Extension points: *Filter*.
- `NodeResourcesFit`: Checks if the node has all the resources that the Pod is
  requesting.

>   Extension points: *Pre-filter*, *Filter*.

Reviewer 2: ```suggestion
  Extension points: *PreFilter*, *Filter*.
```

  Extension points: *Score*.
- `NodeResourcesLeastAllocated`: Favors nodes that have a low allocation of
  resources.
  Extension points: *Score*.
- `VolumeBinding`: Checks if the node has or if it can bind the requested
  volumes.
  Extension points: *Filter*.
- `VolumeRestrictions`: Checks that volumes mounted in the node satisfy
  restrictions which are specific to the volume provider.
  Extension points: *Filter*.
- `VolumeZone`: Checks that volumes requested satisfy any zone requirements they
  might have.
  Extension points: *Filter*.
- `NodeVolumeLimits`: Checks that CSI volume limits can be satisfied for the
  node.
  Extension points: *Filter*.
- `EBSLimits`: Checks that EBS volume limits can be satisfied for the node.
  Extension points: *Filter*.
- `GCEPDLimits`: Checks that GCP-PD volume limits can be satisfied for the node.
  Extension points: *Filter*.
- `AzureDiskLimits`: Checks that Azure disk volume limits can be satisfied for
  the node.
  Extension points: *Filter*.
- `InterPodAffinity`: Implements
  [inter Pod affininity and antiaffinity](/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity).

>   Extension points: *Pre-filter*, *Filter*, *Pre-score*, *Score*.

Reviewer 2: ```suggestion
  Extension points: *PreFilter*, *Filter*, *PreScore*, *Score*.
```

- `PrioritySort`: Provides the default priority based sorting.
  Extension points: *QueueSort*.
- `DefaultBinder`: Provides the default binding mechanism.
  Extension points: *Bind*.

A profile can be specified in a file (via `--config` flag) using the component
config APIs ([`v1alpha1`](https://pkg.go.dev/k8s.io/kube-scheduler@v0.18.0/config/v1alpha1?tab=doc#KubeSchedulerConfiguration)
or [`v1alpha2`](https://pkg.go.dev/k8s.io/kube-scheduler@v0.18.0/config/v1alpha2?tab=doc#KubeSchedulerConfiguration)).

In addition to the default plugins, the following plugins can also be enabled
through these APIs:

- `NodeResourcesMostAllocated`: Favors nodes that have a high allocation of
  resources.
  Extension points: *Score*.
- `RequestedToCapacityRatio`: Favor nodes according to a configured function of
  the allocated resources.
  Extension points: *Score*.
- `NodeResourceLimits`: Favors nodes that satisfy the Pod resource limits.

>   Extension points: *Pre-score*, *Score*.

Reviewer 2: ```suggestion
  Extension points: *PreScore*, *Score*.
```

- `CinderVolume`: Checks that Cinder volume limits can be satisfied for the
  node.
  Extension points: *Filter*.
- `NodeLabel`: Filters and/or scores a node according to the configured labels.
  Extension points: *Filter*, *Score*.
- `ServiceAffinity`: Checks that Pods that belong to a service fit in a set of
  nodes defined by configured labels. It also favors spreading the Pods
  belonging to a service across nodes.

>   Extension points: *Prefilter*, *Filter*, *Score*.

Reviewer 2: ```suggestion
  Extension points: *PreFilter*, *Filter*, *Score*.
```

  
#### Multiple profiles

When using the component config API v1alpha2, a scheduler can be configured to
run more than one profile. Each profile has an associated scheduler name.
Pods that want to be scheduled according to a specific profile can include
the corresponding scheduler name in its `.spec.schedulerName`. If a Pod doesn't
specify a scheduler name, it is scheduled by the profile named

> `default-scheduler`.

Reviewer 2: ```suggestion
the corresponding scheduler name in its `.spec.schedulerName`. Note that `.spec.schedulerName` is set to `default-scheduler` if the Pod doesn't specify one, therefore a profile with this scheduler name should exist to get those pods scheduled.
```

  that don't have Pods for the service already assigned there. The overall outcome is
  that the Service becomes more resilient to a single Node failure.

- `EqualPriority`: Gives an equal weight of one to all nodes.

- `EvenPodsSpreadPriority`: Implements preferred
  [pod topology spread constraints](/docs/concepts/workloads/pods/pod-topology-spread-constraints/).

### Scheduling profiles

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

> Scheduling profiles allow more flexibility when configuring a scheduler.

Reviewer: >  Scheduling profiles allow more flexibility when configuring a scheduler.

If this means “more than the old code”, consider rewording. Ideally, you could keep the same text for the alpha, beta and the GA feature (possibly dropping a bit about “enable feature gate _FooBar_ first”).

Author: I'm happy to just drop the sentence. On another note, there is no feature gate. Users just have to use the v1alpha2 API.

Author: Removed

   about a Pod or the cluster before filtering.
1. **Filter**: These plugins are the equivalent of Predicates in a scheduling
   Policy and are used to filter out nodes that can not run the Pod. Filters
   are called in the configured order.
1. **PreScore**: This is an informational extension point that can be used
   for doing pre-scoring work.
1. **Score**: These plugins provide a score to each node that has passed the
   filtering phase. The scheduler will then select the node with the highest
   weighted scores sum.
1. **Reserve**: This is an informational extension point that notifies plugins
   when resources have being reserved for a given Pod.
1. **Permit**: These plugins can prevent or delay the binding of a Pod.
1. **PreBind**: These plugins perform any work required before a Pod is bound.
1. **Bind**: The plugins bind a Pod to a Node. Bind plugins are called in order
   and once one has done the binding, the remaining plugins are skipped. At
   least one bind plugin is required.
1. **PostBind**: This is an informational extension point that is called after
   a Pod has been bound.
1. **UnReserve**: This is an informational extension point that is called if

>    a Pod is rejected after being reserved and put on hold by a Permit plugin.

Reviewer: How do you feel about putting this level of detail in a page in the [Reference](https://kubernetes.io/docs/reference/) section?

Author: I assume you mean https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/ to be more specific. And that should include both `Scheduling Policies` and `Scheduling Profiles` sections. I think that would be better.

Reviewer: A warning: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/ is [generated](https://kubernetes.io/docs/contribute/generate-ref-docs/kubernetes-components/) documentation.

So not that particular page, as it's not easy to add content to generated docs, but somewhere else in the reference section.

Author: In that case, see kubernetes/kubernetes#88919

Author: As long as the issue above is not resolved, I think adding another reference page would just add confusion about the intent of each file.

Reviewer: I think it's OK to add a page with a URI like https://kubernetes.io/docs/reference/scheduling/plugins/ that lists all the plugins and what they do.

I'd see that as equivalent to the list of admission controllers in https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#what-does-each-admission-controller-do

Author: In that case, I think we can have:

- kubernetes.io/docs/reference/scheduling/policies
- kubernetes.io/docs/reference/scheduling/profiles

Which are linked from kube-scheduler-implementation (which I also plan to rename to `How a node is selected`)

Reviewer: :ok_hand:

Author: Done

  volumes.
  Extension points: *Filter*.
- `VolumeRestrictions`: Checks that volumes mounted in the node satisfy
  restrictions which are specific to the volume provider.
  Extension points: *Filter*.
- `VolumeZone`: Checks that volumes requested satisfy any zone requirements they
  might have.
  Extension points: *Filter*.
- `NodeVolumeLimits`: Checks that CSI volume limits can be satisfied for the
  node.
  Extension points: *Filter*.
- `EBSLimits`: Checks that EBS volume limits can be satisfied for the node.
  Extension points: *Filter*.
- `GCEPDLimits`: Checks that GCP-PD volume limits can be satisfied for the node.
  Extension points: *Filter*.
- `AzureDiskLimits`: Checks that Azure disk volume limits can be satisfied for
  the node.
  Extension points: *Filter*.
- `InterPodAffinity`: Implements
  [inter Pod affininity and antiaffinity](/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity).
  Extension points: *PreFilter*, *Filter*, *PreScore*, *Score*.
- `PrioritySort`: Provides the default priority based sorting.
  Extension points: *QueueSort*.
- `DefaultBinder`: Provides the default binding mechanism.
  Extension points: *Bind*.

> A profile can be specified in a file (via `--config` flag) using the component

Reviewer: ```suggestion
You can specify a profile in a configuration file (via `--config` flag) using the component
```

Author: Done

config APIs ([`v1alpha1`](https://pkg.go.dev/k8s.io/kube-scheduler@v0.18.0/config/v1alpha1?tab=doc#KubeSchedulerConfiguration)
or [`v1alpha2`](https://pkg.go.dev/k8s.io/kube-scheduler@v0.18.0/config/v1alpha2?tab=doc#KubeSchedulerConfiguration)).

In addition to the default plugins, the following plugins can also be enabled

> through these APIs:

Reviewer: How about:
```suggestion
You can also enable the following plugins through those APIs, that
are not enabled by default:
```

Author: Done

  Extension points: *Filter*, *Score*.
- `ServiceAffinity`: Checks that Pods that belong to a service fit in a set of
  nodes defined by configured labels. It also favors spreading the Pods
  belonging to a service across nodes.
  Extension points: *PreFilter*, *Filter*, *Score*.
  
#### Multiple profiles

When using the component config API v1alpha2, a scheduler can be configured to
run more than one profile. Each profile has an associated scheduler name.
Pods that want to be scheduled according to a specific profile can include
the corresponding scheduler name in its `.spec.schedulerName`.

By default, one profile with the scheduler name `default-scheduler` is created.
This profile includes the default plugins described above. When declaring more
than one profile, a unique scheduler name for each of them is required.

If a Pod doesn't specify a scheduler name, kube-apiserver will set it to
`default-scheduler`. Therefore, a profile with this scheduler name should exist
to get those pods scheduled.

{{< note >}}
Scheduling events that are associated to a Pod have the scheduler name defined
for each profile as the ReportingController. Events for leader election use the

> scheduler name of the first profile in the list.

Reviewer 2: ```suggestion
Pod's scheduling events have spec.schedulerName as the ReportingController. Events for leader election use the
scheduler name of the first profile in the list.
```

Author: Done

{{< /note >}}

{{< note >}}

> All the profiles must use the same plugin in the QueueSort extension point and

Reviewer 2: ```suggestion
All profiles must use the same plugin in the QueueSort extension point and
```

Author: Done
